File: Yandex/161.one-edit-distance.java
/**
 * Algorithm:
 * Ensure s is no longer than t. At the first mismatch, equal lengths require
 * replacing one character; unequal lengths require inserting one character into
 * s, so compare the remaining suffixes accordingly.
 *
 * Complexity:
 * Time O(n), space O(1) excluding substring comparison allocation.
 */
class Solution {
    public boolean isOneEditDistance(String s, String t) {
        if (Math.abs(s.length() - t.length()) > 1) {
            return false;
        }
        if (s.length() > t.length()) {
            return isOneEditDistance(t, s);
        }
        for (int i = 0; i < s.length(); i++) {
            if (s.charAt(i) != t.charAt(i)) {
                if (s.length() == t.length()) {
                    return s.substring(i + 1).equals(t.substring(i + 1));
                }
                return s.substring(i).equals(t.substring(i + 1));
            }
        }
        return t.length() - s.length() == 1;
    }
}

File: Google/161.one-edit-distance.py
"""
Approach: Compare strings at the first mismatch and check the only possible edit.
Data structure: indexes only; no DP table is needed because exactly one edit is allowed.
Interview logic: if lengths differ by more than one, impossible. At the first mismatch, equal lengths require replacing one char; unequal lengths require deleting one char from the longer string. If no mismatch occurs, strings are one edit apart only when their lengths differ by one.
Complexity: O(n) time, O(1) space.
Tests and edge cases: identical strings return False; empty vs one char returns True; empty vs empty returns False.
"""
from __future__ import annotations

# @lc code=start
class Solution:
    def isOneEditDistance(self, s: str, t: str) -> bool:
        if abs(len(s) - len(t)) > 1:
            return False
        if len(s) > len(t):
            s, t = t, s
        for i in range(len(s)):
            if s[i] != t[i]:
                if len(s) == len(t):
                    return s[i + 1:] == t[i + 1:]
                return s[i:] == t[i + 1:]
        return len(t) - len(s) == 1
# @lc code=end
